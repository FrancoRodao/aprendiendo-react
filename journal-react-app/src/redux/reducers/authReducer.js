import { types } from "../types/types";

const initialState = {}

export const authReducer = (state = initialState, action) => {
    switch (action.type) {
        case types.authLogin:
            return {
                uid: action.payload.uid,
                name: action.payload.displayname
            };

        case types.authRegister:

            return {
                uid: action.payload.uid,
                name: action.payload.displayname
            };

        case types.authLogout:

            return {}

        default:
            return state;
    }
}