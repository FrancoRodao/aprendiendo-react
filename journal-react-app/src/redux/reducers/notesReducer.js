import { types } from "../types/types";

const initialState = {
    notes: [],
    active: null
}

export const notesReducer = (state = initialState, action) => {
    switch (action.type) {

        case types.notesAdd: 
            return {
                ...state,
                notes: [...state.notes, {
                    id: action.payload.id, 
                    ...action.payload.note
                }]
            }

        case types.notesSetActive:

            return {
                ...state,
                active: {
                    ...action.payload
                }
            }

        case types.notesLoad:
            return {
                ...state,
                notes: [...action.payload]
            }

        case types.notesUpdate:

            return {
                ...state,
                notes: state.notes.map(
                    note => (note.id === action.payload.note.id)
                        ?
                        action.payload.note : note
                )
            }

        case types.notesDelete:
            return {
                ...state,
                notes: state.notes.filter(note => note.id === action.payload.note.id ? null:note),
                active: null
            }
        
        default:
            return state
    }
}