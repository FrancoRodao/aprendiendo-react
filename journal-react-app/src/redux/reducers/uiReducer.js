import { types } from "../types/types";

const initialState = {
    loading: false,
    errors: {
        state: false,
        lastMsg: ""
    }
}

export const uiReducer = (state = initialState, action) => {
    switch (action.type) {
        case types.uiStartLoading:
            return {
                ...state,
                loading: action.payload.loading
            }

        case types.uiFinishLoading:
            return {
                ...state,
                loading: action.payload.loading
            }

        case types.uiError:
            return {
                ...state,
                errors: {
                    //state: true have at least one error
                    state: action.payload.state,
                    // if the message does not arrive, do not change it to undefined
                    lastMsg: (action.payload.lastMsg === '') ? '' : (action.payload.lastMsg || state.errors.lastMsg)
                }
            }

        default:
            return state
    }

}