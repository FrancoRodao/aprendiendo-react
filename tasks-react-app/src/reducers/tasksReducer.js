export const taskReducer = (state, {type, payload})=>{
    switch (type) {
        case "add": 
        
            return [...state, payload]

        case "edit":

            return state.map(task => task.id === payload.id ? payload:task)

        default:
            return state
    }
}