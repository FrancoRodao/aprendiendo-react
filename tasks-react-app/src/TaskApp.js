import React, { useEffect, useReducer } from 'react'
import { taskReducer } from "./reducers/tasksReducer";
import AddTask from './components/AddTask'
import { TaskItem } from './components/TaskItem'
import { useForm } from './hooks/useForm'

const initialState = JSON.parse(localStorage.getItem("tasks")) || []

const TaskApp = () => {

    const [tasks, dispatch] = useReducer(taskReducer, initialState)
    useEffect(() => {
        localStorage.setItem("tasks", JSON.stringify(tasks))
    }, [tasks])

    const [formValues, handleInputChange] = useForm({
        title: "",
        desc: "",
        done: false
    })

    const handleSubmit = (e)=>{
        e.preventDefault()
       dispatch({
           type: "add",
           payload: {
               id: tasks.length + 1,
               ...formValues
           }
       })
    }

    return (
        <>
            <h1>Tasks APP</h1>

            <AddTask handleSubmit={handleSubmit} handleInputChange={handleInputChange} values={formValues} />
            
            {
                tasks.map(task => <TaskItem {...task} key={task.id} dispatch={dispatch} task={task}/>)
            }
        </>
    )
}

export default TaskApp