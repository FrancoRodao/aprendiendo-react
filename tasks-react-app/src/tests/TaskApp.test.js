import React, { useReducer } from 'react'
import { shallow } from "enzyme";
import TaskApp from "../TaskApp";
import { renderHook, act } from '@testing-library/react-hooks'
import { taskReducer } from '../reducers/tasksReducer';

describe('TaskApp component test', () => {

    test('should display a snapshot correctly', () => {
        const wrapper = shallow(<TaskApp/>)
        expect(wrapper).toMatchSnapshot()
    })
    
    test('should add new task | test handleSubmit function', () => {
        const initialState = []
        const {result} = renderHook(() => useReducer(taskReducer, initialState))
        const [tasks, dispatch] = result.current

        const newTask = {
            id: tasks.length + 1,
            title: "new title",
            desc: "new desc",
            done: false
        }

        act(()=>{
            dispatch({
                type: "add",
                payload: {
                    ...newTask
                }
            })
        })

        const [newStateTasks] = result.current
        expect(newStateTasks).toEqual([
            ...initialState,
            newTask
        ])
    })
    
})
