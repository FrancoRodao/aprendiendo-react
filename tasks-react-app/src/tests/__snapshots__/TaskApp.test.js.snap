// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskApp component test should display a snapshot correctly 1`] = `
<Fragment>
  <h1>
    Tasks APP
  </h1>
  <AddTask
    handleInputChange={[Function]}
    handleSubmit={[Function]}
    values={
      Object {
        "desc": "",
        "done": false,
        "title": "",
      }
    }
  />
</Fragment>
`;
